File: bbctl/bbctl/sale.py
# Copyright (C) 2019  Mathias Lohne

# This program is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.

# You should have received a copy of the GNU General Public License along
# with this program.  If not, see <http://www.gnu.org/licenses/>.

from datetime import datetime
from pathlib import Path
import pytz
from typing import *

from .common import *

historyfile = "{}/.bbdebet2/saleshistory.csv".format(str(Path.home()))


def sale(args: List[str] = []):
    executor = CommandExecutor("sale")
    executor.add_command(Command("latest", list_latest, "get the latest sales"))
    executor.add_command(Command("list", list_all, "list all sales"))

    verify_and_run_executor(executor, args)


def list_latest(args: List[str] = []):
    if is_help(args):
        print("List the latest sales. Optional argument: number of sales (default 10).")
        return

    number = int(args[0]) if len(args) >= 1 else 10

    print("{:20} {:12} {:25} {:10} {:10}".format("Time", "User", "Product", "Price", "Earnings"))
    print("-"*(21+13+26+11+10))

    total_sold = 0
    total_earnings = 0

    for sale in parse_csv_file(historyfile)[-number:]:
        utc_dt = datetime.utcfromtimestamp(int(sale[1]))

        formatted_dt = utc_dt.astimezone(pytz.timezone("Europe/Oslo")).strftime("%d.%m - %H:%M:%S")
        formatted_price = "{:.2f} kr".format(float(sale[4]))
        formatted_earnings = "{:.2f} kr".format(float(sale[5]))

        print("{:20} {:12} {:25} {:10} {:10}".format(formatted_dt, sale[2].capitalize(), sale[3][0: min(len(sale[3]),25)], formatted_price, formatted_earnings))

        total_sold += float(sale[4])
        total_earnings += float(sale[5])

    print("-"*(21+13+26+11+10))
    print("Total sold:     {:.2f} kr".format(total_sold))
    print("Total earnings: {:.2f} kr".format(total_earnings))



def list_all(args: List[str] = []):
    if is_help(args):
        print("List all sales this semester.")
        return

    print("{:20} {:12} {:25} {:10} {:10}".format("Time", "User", "Product", "Price", "Earnings"))
    print("-"*(21+13+26+11+10))

    total_sold = 0
    total_earnings = 0

    for sale in parse_csv_file(historyfile):
        utc_dt = datetime.utcfromtimestamp(int(sale[1]))

        formatted_dt = utc_dt.astimezone(pytz.timezone("Europe/Oslo")).strftime("%d.%m - %H:%M:%S")
        formatted_price = "{:.2f} kr".format(float(sale[4]))
        formatted_earnings = "{:.2f} kr".format(float(sale[5]))

        print("{:20} {:12} {:25} {:10} {:10}".format(formatted_dt, sale[2].capitalize(), sale[3][0: min(len(sale[3]),25)], formatted_price, formatted_earnings))

        total_sold += float(sale[4])
        total_earnings += float(sale[5])

    print("-"*(21+13+26+11+10))
    print("Total sold:     {:.2f} kr".format(total_sold))
    print("Total earnings: {:.2f} kr".format(total_earnings))

File: src/com/mathiaslohne/bbdebet2/kernel/core/Sale.java
/*
 * Copyright (C) 2019  Mathias Lohne
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.mathiaslohne.bbdebet2.kernel.core;

import java.util.Date;


public class Sale {

    private static long counter;

    private long id;
    private long timestamp;
    private String userName;
    private String productName;
    private double pricePayed;
    private double earnings;


    Sale(long id, long timestamp, String userName, String productName, double pricePayed, double earnings) {
        this.id = id;
        this.timestamp = timestamp;
        this.userName = userName;
        this.productName = productName;
        this.pricePayed = pricePayed;
        this.earnings = earnings;
        if (id > counter) counter = id + 1;
    }


    public Sale(User user, Product product) {
        timestamp = System.currentTimeMillis() / 1000L;
        id = counter++;

        userName = user.getUserName();
        productName = product.getName();
        pricePayed = product.getSellPrice();
        earnings = product.getMarkup();
    }


    public static long getCounter() {
        return counter;
    }


    public static void setCounter(long counter) {
        Sale.counter = counter;
    }


    public boolean madeByUser(User user) {
        return user.getUserName().equals(userName);
    }


    public long getTimestamp() {
        return timestamp;
    }


    public long getId() {
        return id;
    }


    public String getUserName() {
        return userName;
    }


    public void setUserName(String userName) {
        this.userName = userName;
    }


    public String getProductName() {
        return productName;
    }


    public Product getProduct() {
        return new Product(productName, pricePayed, pricePayed - earnings);
    }


    public double getPricePayed() {
        return pricePayed;
    }


    public double getEarnings() {
        return earnings;
    }


    public String getFormattedTimestamp() {
        return Kernel.dateTimeFormat.format(new Date(timestamp * 1000L));
    }


    @Override
    public String toString() {
        return getFormattedTimestamp() + ":    " + productName + " (" + CurrencyFormatter.format(
            pricePayed) + ")";
    }
}
